File: java/src/main/java/com/java/stack/LinkedStack.java
package com.java.stack;

import com.java.list.Node;

public class LinkedStack<E> implements StackInterface<E> {
	private Node<E> topNode;
	private final E ERROR = null;

	public LinkedStack() {
		topNode = null;
	}

	public static void main(String[] args) {
		LinkedStack<Integer> s = new LinkedStack<>();

		s.push(100);
		s.push(200);
		s.push(600);

		System.out.println(s.pop());
		System.out.println(s.top());

		System.out.println(s.isEmpty());
		s.popAll();

		System.out.println(s.isEmpty());
	}


	@Override
	public void push(E newItem) {
		/*
		newNode.item = x
		newNode.next = topNode (현재의탑)
		topNode = newNode (다시 새로운 탑은 뉴노드)
		 */
		topNode = new Node<>(newItem, topNode);
	}

	@Override
	public E pop() {
		if (isEmpty()) return null;
		Node<E> poppedNode = topNode;
		topNode = topNode.next;
		return poppedNode.item;
	}

	@Override
	public E top() {
		return topNode.item;
	}

	@Override
	public boolean isEmpty() {
		return topNode == null;
	}

	@Override
	public void popAll() {
		topNode = null;
	}
}

File: kotlin/src/main/kotlin/com/kotlin/stack/LinkedStack.kt
package com.kotlin.stack


class LinkedStack : StackInterface {
    private var topNode: Node? = null
    private val error = -12345

    override fun push(newItem: Int) {
        topNode = Node(newItem, topNode) // 현재의 topNode가 새로운 topNode의 next로 연결되도록

    }

    override fun pop(): Int? {
        if (isEmpty()) {
            return error
        } else {
            val poppedNode = topNode
            topNode = topNode?.next
            if (poppedNode != null) {
                return poppedNode.item
            }
        }
        return error
    }

    override fun top(): Int? {
        if (isEmpty()) {
            return error
        }
        return topNode?.item
    }

    override fun isEmpty(): Boolean {
        return topNode == null
    }

    override fun popAll() {
        topNode = null
    }

}